File: Sample.java
// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes

//https://leetcode.com/problems/pascals-triangle/description/


class Solution {
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> triangle=new ArrayList<>();
        List<Integer> frow=new ArrayList<>();
        frow.add(1);
        triangle.add(frow);
        for(int i=1;i<numRows;i++)
        {
            List<Integer> prev_row=triangle.get(i-1);
            List<Integer> row=new ArrayList<>();
            row.add(1);
            for(int j=1;j<i;j++)
            {
                row.add(prev_row.get(j)+prev_row.get(j-1));
            }
            row.add(1);
            triangle.add(row);
        }
        return triangle;
    }
}


// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes

//https://leetcode.com/problems/k-diff-pairs-in-an-array/description/

class Solution {
    public int findPairs(int[] nums, int k) {
        HashMap<Integer,Integer> map= new HashMap<Integer,Integer>();
        int count=0;
        for(int i=0;i<nums.length;i++)
        {
            map.put(nums[i],map.getOrDefault(nums[i],0)+1);
        }
        for(int key:map.keySet())
        {
            if(k==0 && map.get(key)>1)
            {
                count++;
            }
            else if(map.containsKey(key-k) && k!=0)
            {
                count++;
            }
        }
        return count;   
    }
}

File: Sample.py
# Time and space complexicity for each problem are in comments under the function(s)
# Problem 1: Pascal's Traingle

class Solution:

    def generate(self, numRows: int) -> List[List[int]]:

        """
        Description: Given a non-negative integer numRows, generate the first numRows of Pascal's triangle.
        
        Approach:
        - Develop a sub-array for each row with all values = 1 (note the begining and end of each row is 1)
        - Replace su-array values from 2nd index to 2nd last index using Pascal's traigle rule
        
        Time Complexicity: O(n^2)
        Space Complexicity: O(n) # for sub_array creation (In addition to resulting list)
        """
        
        if numRows == 0: return []
        
        result = [] 
        for i in range(numRows):
            sub_list = [1]*(i + 1)
            for j in range(1, i):
                sub_list[j] = result[i-1][j-1] + result[i-1][j]
                
            result.append(sub_list)
            
        return result
        
# Problem 2. K diff pairs in an array

class Solution:
    
    def findPairs(self, nums: List[int], k: int) -> int:
        
        """
        Description: Given an array of integers nums and an integer k, return the number of unique k-diff pairs in the array.
        
        Approach: Use a 2-pass solution
        - First Pass: Create a dictionary with frequencies of each element as value and element as key
        - Second Pass: Count if any "key + target" is present in keys from the dictionary
        - Edge case: for k == 0; create a separate pass
        """
        
        if len(nums) == 0: return 0
        num_record = {}
        pair_count = 0

        for num in nums:
            if num in num_record:
                num_record[num] += 1
            else:
                num_record[num] = 1

        if k != 0:
            for key in num_record.keys():
                if key + k in num_record.keys():
                    pair_count += 1
        else:
            for key in num_record.keys():
                if num_record[key] >= 2:
                    pair_count += 1
                    
        return pair_count

